### PR TITLE
Various optimizations

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

This set out as replacing some `.map` with `.pluck` to avoid building objects just to discard them a second later but ended up containing the following things:

- refactoring of `ReactionsController#index` plus some improved tests
- refactoring of `BufferedArticlesController`
- addition of `Reaction.positive` to replace the various `where("points > ?", 0)`
- query optimization for `Reaction.count_for_article` to go from 3 queries to 1
- query optimization for `MailchimpBot#manage_tag_moderator_list` to go from building objects in memory to ask the DB for the values

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

### CODE REVIEW

File: app/controllers/buffered_articles_controller.rb [hunk 1]
```diff
@@ -1,20 +1,20 @@
...
+  private
+
+  def buffered_articles_urls
+    relation = if Rails.env.production?
+                 Article.where("last_buffered > ? OR published_at > ?", 24.hours.ago, 20.minutes.ago)
+               else
+                 Article.all
+               end
+
+    paths = relation.pluck(:path)
```
Comment (rhymes) [inline app/controllers/buffered_articles_controller.rb]: we don't need to build Article objects in memory if we only need the `path`

File: app/controllers/reactions_controller.rb [hunk 2]
```diff
@@ -4,37 +4,49 @@ class ReactionsController < ApplicationController
...
-        {
-          current_user: { id: session_current_user_id },
-          positive_reaction_counts: reaction_counts,
-          reactions: reactions
-        }.to_json
+      comments = Comment.
+        where(commentable_id: params[:commentable_id], commentable_type: params[:commentable_type]).
+        select(%i[id positive_reactions_count])
+
+      # iterating only once
```
Comment (rhymes) [inline app/controllers/reactions_controller.rb]: here we were iterating on comments objects twice even though we could iterate once :D

File: app/controllers/reactions_controller.rb [hunk 3]
```diff
@@ -79,8 +91,7 @@ def create
 
   def cached_user_positive_reactions(user)
     Rails.cache.fetch("cached_user_reactions-#{user.id}-#{user.updated_at}", expires_in: 24.hours) do
-      Reaction.where(user_id: user.id).
-        where("points > ?", 0)
+      user.reactions.positive
```
Comment (rhymes) [inline app/controllers/reactions_controller.rb]: Rails FTW

File: app/labor/mailchimp_bot.rb [hunk 4]
```diff
@@ -76,8 +76,12 @@ def manage_tag_moderator_list
     return false unless user.tag_moderator?
 
     success = false
-    tags = user.roles.where(name: "tag_moderator").map { |tag| Tag.find(tag.resource_id).name }
+
+    tag_ids = user.roles.where(name: "tag_moderator").pluck(:resource_id)
+    tag_names = Tag.where(id: tag_ids).pluck(:name)
```
Comment (rhymes) [inline app/labor/mailchimp_bot.rb]: instead of loading roles and tag objects, we only load arrays

File: app/models/reaction.rb [hunk 5]
```diff
@@ -48,8 +50,10 @@ class << self
     def count_for_article(id)
       Rails.cache.fetch("count_for_reactable-Article-#{id}", expires_in: 1.hour) do
         reactions = Reaction.where(reactable_id: id, reactable_type: "Article")
+        counts = reactions.group(:category).count
+
         %w[like readinglist unicorn].map do |type|
-          { category: type, count: reactions.where(category: type).size }
+          { category: type, count: counts.fetch(type, 0) }
```
Comment (rhymes) [inline app/models/reaction.rb]: here we were sending 3 counting queries to the DB instead of letting the DB use "group by count" once

File: app/controllers/buffered_articles_controller.rb [hunk 6]
```diff
@@ -1,20 +1,20 @@
...
-        select(:path).
-        map { |a| "https://#{ApplicationConfig['APP_DOMAIN']}#{a.path}" }
-    else
-      Article.all.map { |a| "https://#{ApplicationConfig['APP_DOMAIN']}#{a.path}" }
-    end
+  private
+
+  def buffered_articles_urls
+    relation = if Rails.env.production?
+                 Article.where("last_buffered > ? OR published_at > ?", 24.hours.ago, 20.minutes.ago)
```
Comment (citizen428) [inline app/controllers/buffered_articles_controller.rb]: I'm fine leaving at as is, but generally when AR supports an operation directly, I tend to use it:

```ruby
Article.where("last_buffered > ?", 24.hours.ago).or(Article.where("published_at > ?", 20.minutes.ago))
```

The generated SQL is almost identical, the AR version just adds explicit parenthesis around the two conditions.

File: app/controllers/buffered_articles_controller.rb [hunk 7]
```diff
@@ -1,20 +1,20 @@
...
+               else
+                 Article.all
+               end
+
+    paths = relation.pluck(:path)
+    paths.map { |path| "https://#{ApplicationConfig['APP_DOMAIN']}#{path}" }
```
Comment (citizen428) [inline app/controllers/buffered_articles_controller.rb]: I wonder if we should add a helper for this at one point. A quick search shows various places throughout the app where we manually build paths like this.

Obviously not part of this PR, just thinking out loud.

File: app/controllers/reactions_controller.rb [hunk 8]
```diff
@@ -4,37 +4,49 @@ class ReactionsController < ApplicationController
...
+
     if params[:article_id]
       id = params[:article_id]
-      reactions = if session_current_user_id.present?
-                    Reaction.where(reactable_id: id,
-                                   reactable_type: "Article",
-                                   user_id: session_current_user_id).
-                      where("points > ?", 0)
+
+      reactions = if session_current_user_id
```
Comment (mstruve) [inline app/controllers/reactions_controller.rb]: Rather than doing this check later in the action, can we move this check to the very top as a guard clause and then skip all of the processing that we do below if we dont have it. I noticed we also check AFTER loading a bunch of comments.

File: app/controllers/buffered_articles_controller.rb (hunk 6 above)
Comment (rhymes) [reply app/controllers/buffered_articles_controller.rb]: Definitely +1 on `.or`, the code base predates the addition of `.or()` and I didn't notice while reviewing it :) I'm going to change it

File: app/controllers/buffered_articles_controller.rb (hunk 7 above)
Comment (rhymes) [reply app/controllers/buffered_articles_controller.rb]: Agreed. There's also a `current_state_path` in the article decorator that doesn't feel right

File: app/controllers/reactions_controller.rb (hunk 8 above)
Comment (rhymes) [reply app/controllers/reactions_controller.rb]: Mmm I'm not sure we can, we load the comments because we need to compute `positive_reaction_counts` and that's needed for both signed in and non signed in users. The only thing we don't need in case the user is not signed in is the `comment_ids` array, I can move that

File: spec/requests/reactions_spec.rb [hunk 9]
```diff
@@ -50,19 +74,28 @@
...
+        result = response.parsed_body
+
+        expect(result["current_user"]).to eq("id" => user.id)
+        expect(result["positive_reaction_counts"]).to eq([{ "id" => article.comments.last.id, "count" => 1 }])
+        expect(result["reactions"].to_json).to eq(user.reactions.where(reactable: comment).to_json)
       end
 
       it "does not set surrogate key headers" do
-        expect(response.headers["Surrogate-Key"]).to eq(nil)
+        expect(response.headers["surrogate-key"]).to be_nil
```
Comment (mstruve) [inline spec/requests/reactions_spec.rb]: good catch

File: app/controllers/buffered_articles_controller.rb (hunk 7 above)
Comment (citizen428) [reply app/controllers/buffered_articles_controller.rb]: Cool, I added that to my personal todo list.
